### ISSUE TITLE
Get-ToolPath cmdlet is no longer recognised after agent update to 199 on hosted? 

### ISSUE THREAD

**Issue by centur** [Microsoft/vsts-tasks#1710]
This issue is a follow-up of a short [twitter discussion here](https://mobile.twitter.com/centur/status/730201969827053569)
Here is the context of the issue:

Initially, we tried to use few tasks for cloud artefact deployment, and they didn't work well. We hadn't got much success with AZCopy task too, but after some PowerShell fiddling, we got AzCopy tool working in a regular "Azure PowerShell " script.

To discover AzCopy location I was using snippet from one of other tasks I found in the repo:
`$azcopy = Get-ToolPath -Name "AzCopy/AzCopy.exe"`

This script was working fine on Monday 09 May, but started to fail next day, with no obvious reason - we haven't touched our build scripts for a long time. The screenshot in the tweet shows the error message. The only difference between failed and successful builds I found after studying logs, is that agent version on the hosted build changed from 1.98.1 to 1.99

I tried to get working the code above using Azure Powershell task and just a regular Powershell task with same results:

>The 'Get-ToolPath' command was found in the module 'Microsoft.TeamFoundation.DistributedTask.Task.Internal', but the module could not be loaded. For more information, run 'Import-Module Microsoft.TeamFoundation.DistributedTask.Task.Internal'.

I'm not sure if this problem is global, it seems that it may affect only our assigned Hosted Agent instance, but we can't really do anything with it - we can't control Hosted Agent instance. Also, it's a bit strange that previously working CI pipeline started to break without any external changes.

PS: I managed to work around the problem using standard AzCopy task ( this time it started to work out of the box). I also saw that AzCopy Task itself is using similar code to detect AzCopy binary location, and it's working fine now on the same agent. Which makes me think that the problem in the Azure Powershell and Powershell tasks or their contexts\modules, not in the breakage of a particular Agent instance.

**Comment by ericsciple** [comment]
I see. The Microsoft.TeamFoundation.DistributedTask.Task.Internal module is for custom tasks, not ad-hoc scripts. It happened to work before due to the way the old implementation of the task was written. You're script before was relying on an internal implementation detail of how the task was implemented.

On a related note, extra precaution has been taken in the new implementation of the task to prevent folks from relying on internal implementation details of the task. Now all imported task SDK functions are removed before calling the ad-hoc user supplied script. Additionally, the new implementation uses a completely different task SDK altogether - which doesn't have the Get-Tool command at all.

We have some changes coming - no hard timeframe at the moment - to increase the discoverability of tools for custom tasks. Depending on how that work plays out, it may be able to be leveraged from ad-hoc scripts as well.

It sounds like you're unblocked by the AzCopy task. I'm going to go ahead and close this issue. To summarize, the issue you ran into is by design. It was a mistake that the custom task SDK was previously accessible to ad-hoc scripts. That led you down a path that caused the disruption when we switched the underlying execution handler for that task.

**Comment by centur** [comment]
well, previously there was no documentation on what is the right way to discover AzCopy, neither working AzCopy task, so we have to solve the problem in at least some way. 

What is the right way to discover hosted agent tools ?

**Comment by bryanmacfarlane** [comment]
Typically, tools for consumption are added to the PATH (use where - for example node installers do this) or some have well known env vars (maven has M2_PATH, java has JAVA_HOME etc..).  Typically we don't invent a way, we install the tools on the image and let the tool.

Eric, can you check with CJ tomorrow and see if this is on the image anywhere (as opposed to relying on our internal copy in the agent)?

**Comment by ericsciple** [comment]
It sounds like @centur has already found a workaround specific to the azcopy.exe issue - switched to the AzCopy task which wasn't previously available.

I went ahead and ran a hosted build. `where` doesn't find it, but it appears to be in the below locations. As suggested, I'll talk to CJ tomorrow to see what he knows about the reliability of those locations.
```
c:\Program Files (x86)\Microsoft SDKs\Azure\AzCopy\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\Extensions\3uogi5ex.psm\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\Extensions\4bdgrm5s.z0a\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\Extensions\qdqrx5aq.vq4\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\Extensions\usxx1hq1.0ge\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\Extensions\0w52fl0j.zfn\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\Extensions\al0ndy0z.44w\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\Extensions\c50azl4x.cly\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
c:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\Extensions\n4zz5qds.3st\~PC\ProjectTemplates\DeploymentProject\1033\DeploymentProject.zip\Tools\AzCopy.exe
```

**Comment by centur** [comment]
Yes, as I said build is working fine with AzCopy task.

Still keen to know what binary tools and modules we can use in ad-hoc scripts and what is the right way to discover them .
`Get-ToolPath "SomeWellKnownToolName"` seems pretty valid way, would be nice to have official replacement for it.

**Comment by ericsciple** [comment]
I just talked with CJ, it sounds like the hosted image will continue to deliver the Azure SDK. So for this specific tool, `AzCopy.exe`, it sounds like `c:\Program Files (x86)\Microsoft SDKs\Azure\AzCopy\AzCopy.exe` will continue to be a pretty reliable location.

I'll follow-up with @bryanmacfarlane regarding how to find tools in general. This one, specifically, is not in the PATH (appears the installer doesn't add it).

**Comment by ericsciple** [comment]
For tools where the installer doesn't add it to the PATH, I think your best bet would be to look for it in well-known locations (e.g. in this case Azure SDK installation directory).

I'll go ahead an close this issue, since Get-ToolPath (previously delivered by the module Microsoft.TeamFoundation.DistributedTask.Task.Internal) was internal detail of the task implementation, that happened to work before from your ad hoc Azure PowerShell script. The cmdlet was intended only for internal usage by the in-the-box tasks, and was only used to find 4 different tools. The in-the-box tasks away from this module.

I see your point that it would be nice for ad hoc script to have a way to resolve well known tools, but a PowerShell cmdlet cannot be solution. It won't work for any other type of ad hoc script (batch, exe, shell script, etc).

**Comment by centur** [comment]
We have to jump back to ad-hoc scripts when something is broken or not yet exists. So unless you want to build a completely closed system - then it's reasonable to prevent users from finding some ways around the existing problems. 

Otherwise - ad-hoc Powershell script is exactly The Solution - the users last resort to do something they want/need but it wasn't yet designed in the system, or will never be in. And having a friendly basic scripting environment is a good starting point to do this.

Having paths hard-coded usually means - to rely on the default installations, paths and SDKs. It's obviously a way to go sometimes (not a good practice though), but it lacks any discover-ability - user has to know that this tool belongs to specific SDK, of the specific version and that SDK by default installed in particular location and a given agent... 
Well, when there is no nicer way to go around - I'd rather check in my own binaries in the repo and will have a full control there, instead of rely on some hard-coded paths and non-guaranteed promises. Some others may just change a build system and never get back to such problems again
